onnxruntime/core/graph/graph.cc

```diff
@@ -2781,39 +2852,133 @@ Status Graph::InlineFunction(Node& node) {
   for (auto output_edge : output_edges) {
     RemoveEdge(node.Index(), output_edge.GetNode().Index(), output_edge.GetSrcArgIndex(), output_edge.GetDstArgIndex());
   }
+  std::unordered_map<std::string, NodeArg*> remap_input_output;
+  if (node.MutableInputDefs().size() != subgraph.GetInputsIncludingInitializers().size())
+    return Status(ONNXRUNTIME, FAIL, "Node " + node.Name() + "'s number of inputs is different from function body graph's number of input.");
+  for (size_t i = 0; i < subgraph.GetInputsIncludingInitializers().size(); ++i) {
+    auto* input = subgraph.GetInputsIncludingInitializers()[i];
+    if (input->Name() != node.MutableInputDefs()[i]->Name())
+      remap_input_output[input->Name()] = node.MutableInputDefs()[i];
+  }
+
+  ORT_ENFORCE(node.MutableOutputDefs().size() == subgraph.GetOutputs().size(),
+              "Node ", node.Name(), "'s number of outputs is different from function body graph's number of outputs.");
+  for (size_t i = 0; i < subgraph.GetOutputs().size(); ++i) {
+    auto* output = subgraph.GetOutputs()[i];
+    if (output->Name() != node.MutableOutputDefs()[i]->Name())
+      remap_input_output[output->Name()] = node.MutableOutputDefs()[i];
+  }
+
   RemoveNode(node.Index());
   for (const auto& subgraph_node : subgraph.Nodes()) {
-    AddNode(subgraph_node);
+    if (subgraph_node.OpType() == kConstant) {
+      // Copy constant nodes _value to name_to_initial_tensor_
```

Comment:
Is this needed? Usually constant nodes are moved to initializers in Graph::Graph(...).

If it's needed it should use the same logic as the Graph ctor to support opset 12 correctly as the constant value can now come from many different attributes. Just checking 'value' is not longer enough.

Reply:
@souptc can you comment on this?

---
In reply to: [414374516](https://github.com/microsoft/onnxruntime/pull/3681#discussion_r414374516) [](ancestors = 414374516)

Reply:
this is inline the function body (a subgraph) into the main graph, so we need to move the constant value into main graph's initializer manually. at least at that time (last year) the subgraph still contains constant node, but i didn't check that later.

The constant defined in function body (using onnx's utility function) are from 'value', so i think it is fine at this moment, as the inline function is not that common scenario. Could we put a TODO here and resolve it later if people really create function body by using constant in another way?